File: python/clock/clock.py
# coding=utf-8
from __future__ import unicode_literals

__author__ = 'Cedric Zhuang'


class Clock(object):
    def __init__(self, hour, minute):
        self.hour = hour
        self.minute = minute

    def __str__(self):
        self.hour = (self.hour + self.minute // 60) % 24
        self.minute %= 60
        return '{:02}:{:02}'.format(self.hour, self.minute)

    def __eq__(self, other):
        return str(self) == str(other)

    def add(self, minute):
        self.minute += minute
        return self

File: scala/clock/src/main/scala/Clock.scala
/**
  *
  * @author Cedric Zhuang
  *
  */
object Clock {
  def apply(m: Int): Clock = new Clock(0, m)
}

case class Clock(h: Int, m: Int) {
  private lazy val totalMinutes = Math.floorMod(h * 60 + m, 24 * 60)

  lazy val hour: Int = totalMinutes / 60

  lazy val minute: Int = totalMinutes % 60

  def +(other: Clock): Clock = Clock(h + other.h, m + other.m)

  def -(other: Clock): Clock = Clock(h - other.h, m - other.m)

  override def toString: String = f"$hour%02d:$minute%02d"

  override def equals(obj: scala.Any): Boolean =
    obj.isInstanceOf[Clock] && obj.toString == toString
}
